python/agents/Plumber-Data-Engineering-Assistant/plumber_agent/sub_agents/dataflow_agent/tools/dataflow_template_tools.py

```diff
@@ -0,0 +1,317 @@
+       gcs_staging_location: str,
+       custom_gcs_path: str = ""
+   ) -> dict:
+    """
+    Submits a job to Dataflow using either a classic or a Flex template.
+
+    It validates user-provided parameters against the template definition unless a custom GCS path
+    is provided, in which case validation is skipped. It then constructs and executes the
+    appropriate `gcloud` command to launch the job.
+
+    Args:
+        job_name (str): The name to assign to the Dataflow job.
+        input_params (str): A JSON string of the parameters to pass to the template.
+        template_params (str): A JSON string containing the template's definition, including its GCS path and parameter requirements.
+        gcp_project (str): The Google Cloud project ID where the job will run.
+        region (str): The GCP region for the Dataflow job.
+        gcs_staging_location (str): The GCS path for staging temporary files.
+        custom_gcs_path (str, optional): A direct GCS path to a template file. If provided,
+                                         parameter validation is skipped. Defaults to "".
+
+    Returns:
+        dict: A dictionary containing the submission status, stdout from the command, and a comment.
+    """
+    try:
+        user_input_dict = json.loads(input_params)
+        template_definition_dict = json.loads(template_params)
+
+        # Handle cases where the template definition might be wrapped in a list.
+        if isinstance(template_definition_dict, list) and len(template_definition_dict) > 0:
+            template_definition_dict = template_definition_dict[0]
+        
+        template_gcs_path = custom_gcs_path if custom_gcs_path else template_definition_dict.get("template_gcs_path")
+        
+        if not template_gcs_path:
+            return {
+                'status': 'failed',
+                'comment': 'Job could not be submitted because the "template_gcs_path" key was not found.'
+            }
+        
+        # --- MODIFIED LOGIC ---
+        # If a custom GCS path is NOT provided, validate parameters against the definition.
+        # If a custom path IS provided, skip validation.
+        if not custom_gcs_path:
+            if "params" not in template_definition_dict:
+                return {
+                    'status': 'failed',
+                    'comment': 'Job could not be submitted because the template definition is missing the "params" key.'
+                }
+            
+            param_validation_result = validate_input_params(
+                template_definition=template_definition_dict["params"],
+                user_inputs=user_input_dict
+            )
+            if param_validation_result['validation_result'] != 'success':
+                return {
+                    'status': 'failed',
+                    'comment': f'Job could not be submitted due to a parameter validation error: {param_validation_result["comment"]}'
+                }
+
+        delemeter = "###"
```

Comment:
delimiter ?

Reply:
This is a variable which is used while using gcloud command to change default delimiter of gcloud command - which was clashing with upstream command.